testlib/src/main/kotlin/io/spine/protodata/testing/PipelineSetup.kt

```diff
@@ -0,0 +1,276 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.testing
+
+import com.google.common.annotations.VisibleForTesting
+import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest
+import io.spine.io.Resource
+import io.spine.io.ResourceDirectory
+import io.spine.protodata.backend.Pipeline
+import io.spine.protodata.plugin.Plugin
+import io.spine.protodata.renderer.SourceFileSet
+import io.spine.protodata.settings.SettingsDirectory
+import io.spine.protodata.testing.PipelineSetup.Companion.byResources
+import io.spine.reflect.CallerFinder.findCallerOf
+import io.spine.tools.code.Java
+import io.spine.tools.code.Kotlin
+import io.spine.tools.code.Language
+import io.spine.tools.code.Protobuf
+import io.spine.tools.code.TypeScript
+import io.spine.tools.prototap.Names.PROTOC_PLUGIN_NAME
+import io.spine.tools.prototap.Paths.CODE_GENERATOR_REQUEST_FILE
+import java.nio.file.Path
+
+/**
+ * Creates a [Pipeline] for testing given ProtoData [plugins].
+ *
+ * This class simulates the first step of the code generation process
+ * performed by `protoc` compiler. Since `protoc` is a stable and predictable piece of
+ * software, we do not need to go through the "vanilla" code generation process when we
+ * test ProtoData plugins.
+ *
+ * Instead of running the whole code generation process in a Gradle build with ProtoData Gradle
+ * Plugin applied over and over again, we can generate the code and related binary data like
+ * [CodeGeneratorRequest] or [FileDescriptorSet][com.google.protobuf.DescriptorProtos.FileDescriptorSet]
+ * using `protoc` and then use its output in tests.
+ *
+ * A convenient way for capturing the generated "vanilla" code and associated files is
+ * using [ProtoTap Gradle Plugin](https://github.com/SpineEventEngine/ProtoTap).
+ * This is the recommended way for working with `PipelineSetup` which provides most of the input
+ * for creation of a [Pipeline] automatically. Configuration steps associated with this approach are
+ * described below.
+ *
+ * The class also allows fine-tuned way of working using its constructor directly for the usage
+ * scenarios when generated code and [CodeGeneratorRequest] are already available or generated via
+ * a custom procedure.
+ *
+ * ### Creating `PipelineSetup` with the help of ProtoTap
+ * 
+ * Here are the steps for creating an instance of `PipelineSetup` in your tests.
+ *
+ * #### 1. Add [`java-test-fixtures`](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures) plugin
+ * The plugin will create the `testFixtures` source set in which we will put proto files.
+ *
+ * Alternatively, you can put proto files under the `test` source set, but with `testFixtures`
+ * it is a bit more neat.
+ *
+ * ### 2. Add [Protobuf Gradle Plugin](https://github.com/google/protobuf-gradle-plugin)
```

Comment:
Please keep the header level consistent. Here and below it should be `####`.

Reply:
I've moved them all one level up, making nesting consistent.